### ISSUE TITLE
How to throw exceptions in AQL queries?

### ISSUE THREAD

**Issue by konsultaner** [arangodb/arangodb#4583]
I currently work on a huge project that includes a lot of 100+ lined AQL queries. Since the AQL queries are executed in the database, I cannot throw any exceptions if conditions should cause one. Is there a way to use UDF (not causing to much trouble) or pure AQL to throw Exceptions and roleback if any condition is not acceptable? That would help a lot to provide better user feedback, i.e. 

```aql
FOR u IN users
  FILTER u.username = @username
  UPDATE { _key: u._key, loginCount: u.loginCount++ } 
  // field,assertion,errorCode,message,rollback
  ASSERT(u.active,true,1001,"User is not active. This user should not be able to log in!",true)
IN users
```

**Comment by konsultaner** [comment]
Nice to see that you seem to like my idea. I went through the fix and saw that you are to implement the assert and warn with only [2 paramters](https://github.com/arangodb/arangodb/pull/4584/commits/d04d18d25648e4df70935c298b880bfcdf1bc9d1#diff-ea66e162033c2a25bda75cf899e9d067R191). I usually develop in english but have multilingual frontends. Since the idea came up to provide better user feedback, it would be awesome to also have a third parameter to provide an error code. This way I could i18n the correct error message to the user.

**Comment by ObiWahn** [comment]
Hi @konsultaner,
we decided against a 3rd parameter. We did this because users are able to encode anything in the string. You could for example use the first few bytes or anything up to a separator (❤) to encode the code you need. Furthermore it would be hard for us to map the user provided error code. How should it be returned? The user provided code could overwrite our own internal code or the result would need to get an extra field. All solutions that propagate the code outside the message seem to lead to inconsistencies in ArangoDB.

Please close this issue if the PR is sufficient for you!

**Comment by konsultaner** [comment]
@ObiWahn Makes sense!

**Comment by konsultaner** [comment]
@ObiWahn doesn't `ASSERT()` and `WARN()` make `FAIL()` obsolete?

**Comment by konsultaner** [comment]
@jsteemann In the docs the example looks like this:

    FOR i IN 1..3 RETURN FILTER ASSERT(i>0, "i is not greater 0") RETURN i
    FOR i IN 1..3 RETURN FILTER WARN(i<2, "i is not smaller 2") RETURN i

but shouldn't it be like this:

    FOR i IN 1..3 FILTER ASSERT(i>0, "i is not greater 0") RETURN i
    FOR i IN 1..3 FILTER WARN(i<2, "i is not smaller 2") RETURN i

**Comment by jsteemann** [comment]
Yes, seems the examples are wrong.

**Comment by jsteemann** [comment]
Fixed via 9046388722dda8fe795a64621f0a07cbf9a51d28.

**Comment by Simran-B** [comment]
@konsultaner 

> doesn't ASSERT() and WARN() make FAIL() obsolete?

Don't think so. `FAIL()` was never intended to be used by users. It is used in unit tests to test the failure case.
